### ISSUE TITLE
Connect to running SAS process

### ISSUE THREAD

**Issue by MeetQingLiu** [sassoftware/saspy#34]
Hello,
I learn that the same Ipython kernel could be connected by two or more processes.
My question is, Is it possible to connect to a running SAS process such as SAS Display Manager or SAS EG using SASPy?
I mean working in SAS DM and Jupyter which running a same process at the same time.
Thanks.

**Comment by tomweber-sas** [comment]
Thank you, but I'm afraid no, saspy cannot connect to an existing SAS session. You are correct that you can connect to multiple (new) SAS sessions from a single saspy session, like a notebook.

**Comment by MeetQingLiu** [comment]
@tomweber-sas ,Thanks. Did you mean that eveytime I submit `sas = saspy.SASsession(cfgname='config')` will start a new SAS session?

**Comment by tomweber-sas** [comment]
In general, yes. Invoking SASsession will make a new connection to whatever configuration you specify. Saying that, I will say that there is the potential for returning the existing connection that already exists if you are assigning the same object to a session:
sas1 = SASsession()
sas1 = SASsession()

as opposed to using different object:

sas1 = SASsession()
sas2 = SASsession()

In the SASsession code, I check to see if the object I am handed (sas1 or sas2 in the above examples) already has a connection established, and if so, I just return the object as is. When that object is deleted, by python, it drives my code to disconnect from, and shut down SAS. In the case of reassigning the same objec (sas1 -> sas1) I believe I have seen different python behavior across some versions. I used to see the object get deleted and then recreated, meaning I would see the disconnect of the original SAS session and then connect to a new one. Currently, I don't see the delete for this object until the python process is shuting down (I see this in the Jupyter notebook), But I also don't get the same actual object with the connection already established. So, a new SAS session is established and when the notebook is stopped, then I see both objects get deleted and their SAS sessions shut down.

I used to see it overlap, but it just depends on python garbage collection and other things I don't control. So, yes, each SASsession() invocation creates a new connection to a new SAS session.
